bugfix: Override subcharts with null values

ryanhockstad opened the pull request:
This PR closes #12469  and closes #12488

Helm should allow users to not only override default values, but also completely remove any default values by setting a config to `null`. 

This works fine for regular charts, but default values within sub-charts cannot be `null`-ed.  The linked issue has a good example of this created by user "naemono." 

The reason this issue is happening is because the `coalesce` function goes over sub-chart values that are defined in a values file or with a `--set` flag twice due to the logic [here](https://github.com/helm/helm/blob/d37e2e9097f9715d4d184b2e3cc313588f106030/pkg/chartutil/coalesce.go#L100-L101).  `merge` is always set to `false` in this context, and the first time `coalesce` gets called, the null value gets removed during the `coalesceTablesFullKey` function [here.](https://github.com/helm/helm/blob/d37e2e9097f9715d4d184b2e3cc313588f106030/pkg/chartutil/coalesce.go#L279) This is fine for regular chart values, but for sub-chart values, the `coalesceDeps` function runs the logic again for every value in the subChart, so we end right back to the spot in the previous link.  But the config option that was explicitly set to null was already removed, so we end up at [this point](https://github.com/helm/helm/blob/d37e2e9097f9715d4d184b2e3cc313588f106030/pkg/chartutil/coalesce.go#L281) two lines down, which overwrites the user defined `null` config the sub-chart default value.

This problem only occurs because when the `coalesce` function is run while `merge` is `false`.  

To prevent this bug from removing sub-chart config values that are explicitly set to `null` and therefore getting overwritten when `coalesceDeps` runs through the sub-chart values a second time, I added logic to check if the key was the name of a sub-chart, and if so setting `merge` to `false` so the null config option is not deleted.
 


**Special notes for your reviewer**:
Follow-up if this PR is accepted may be to add a unit-test to confirm sub-chart nulls aren't overwritten.

ryanhockstad committed:
merge null child chart objects

Signed-off-by: Ryan Hockstad <ryanhockstad@gmail.com>

messiahUA commented:
When this could be merged?

AndrewFarley commented:
Please merge/release ASAP.  All builds from 3.13 onwards are "effectively" broken for many of my clients because of this.

ssttehrani commented:
We are also blocked and we can not upgrade our ArgoCD instances above 2.9.1

aabouzaid commented:
I hope to see this in the next Helm release, there are many workarounds used to overcome this bug.

duplabe commented:
I think this is also related https://github.com/helm/helm/issues/12991

AndrewFarley commented:
Beep boop we need this, boop beep

AndrewFarley commented:
Ahh, alas, another week goes by, and yet more DevOps engineers are fraught with this unnecessary pain in having to continue to use such an old version of Helm without this bug.  Beep boop, someone merge and release this, boop beep.

AndrewFarley commented:
Another week nudge on this.  Sorry to be the annoying jerk, but this is important.

Tawmu commented:
Sorry to go against etiquette here with comment spam but we're also affected by this bug, and with this PR waiting for review for almost 3 months now I think some of us might be losing hope :-)

pisto commented:
Just collecting all the duplicate bugs about this over the years, for fun I guess...

#12991 #12741 #12637 #12594 #12522 #12511 #12490 #12488 #12469 #12441 #12417 #11567 #9806 #9804 #9696 #9136 #9027 #6277 #5184

I'm so happy that the helm devs keep this bug alive so that I can charge my clients *hours* in working around it. Thanks! Tonight I eat steak thanks to you!

On a more serious note, who can we ping about this? I see @mattfarina worked on #12480 , sorry for the rude ping but maybe you were at least aware of the issue at some point in time.

MrBlaise commented:
I'd like to join the ping party, even though I know it won't help and normally I'd consider it rude as well, but this is getting ridiculous. Even if this is merged it will be some time until tools like argo-cd picks it up so that things can finally go back to the way they were.

Anyone with merge rights on this repo please take a look, lots of people would appreciate it ❤️

AndrewFarley commented:
Hmm, can we just start aggressively @'d owners and/or higher tier contributors on this repo?  I feel like maybe this thread has been muted

sylwit commented:
This PR fixes a really inconvenient bug and is highly expected to be merge. 
@joejulian do you know who we should ping to have a look at it or if this PR isn't ready can we have some feedback please ?
Thank you ❤️

joejulian commented:
I really want @mattfarina to look at this first. He recently dove deep into this and has the most context.

dannyg28 commented:
Is there any update on this? Like others we're blocked on Argo version updates until this defect is resolved

rptaylor commented:
There are Helm developer meetings every Thursday I believe: https://docs.google.com/document/d/1d-6xJEx0C78csIYSPKJzRPeWaHG_8W1Hjl72OJggwdc/edit?pli=1#heading=h.cm5b1kdljh4m
Maybe an interested contributor could join to discuss the next steps for this PR (be nice!)

arya-harness commented:
We would really like this change as well. Is it possible to know which older version of helm didnt have this issue?

c0deaddict commented:
Also waiting for this. I want to override some default cpu limits of a subchart in ArgoCD, but can't..

swagnerfaw commented:
Is this still broken? Whats the reason this is not fixed for such a long time?

shastaxc commented:
Please?

AndrewFarley commented:
This issue going completely ignored so long makes me want to write a cronjob which just auto-reposts in here every day for infinity until this gets merged.  There's no reason for this, this is just causing a bunch of us needless pain and suffering.  Arguably, work, but, painful work.  

Please someone review/merge/fix this.  I'm going to create a new issue right now to reference this issue in this repo, in case this issue is on a ignore/unsubscribe list.

TylerHelmuth commented:
The [OpenTelemetry Helm Charts](https://github.com/open-telemetry/opentelemetry-helm-charts/issues/1160) would also love this feature.  I could've sworn this was fixed in `3.13`, but it appears not.

pisto commented:
Alright I'll bite and join the weekly meeting, if I am not busy. I am no longer affected by this bug (all charts I manage are custom), nor I am confident with the PR code itself, so I'd highly appreciate the presence of @ryanhockstad . Otherwise, I'll just point to this PR and explain the situation.

pisto commented:
I have a reproducible testcase btw https://github.com/pisto/helm-subchart-null-bug

ChieloNewctle commented:
I encountered the same problem when attempting to unset `seccompProfile` in the subchart `kong` of `kubernetes-dashboard`.
Sorry for anothing weekly ping. But this fix is really needed. Pelase. @mattfarina

MrBlaise commented:
> Alright I'll bite and join the weekly meeting, if I am not busy. I am no longer affected by this bug (all charts I manage are custom), nor I am confident with the PR code itself, so I'd highly appreciate the presence of @ryanhockstad . Otherwise, I'll just point to this PR and explain the situation.

Did you have a chance to join the meeting?

Plus another PING for me. Please for the love of god, should we fork helm is this project abandoned? I am obviously joking but this PR sure makes it look like a dead / amateur project.

messiahUA commented:
just a reminder that we can celebrate 5 months since this bug was fixed in the PR

gionn commented:
@mattfarina can we get a feedback on this from a maintainer, please?

pierluigilenoci commented:
Let's see if anything moves... 💣 

@joejulian
@jdolitsky
@marckhouzam
@mattfarina
@sabre1041 
@scottrigby 
@technosophos 

Please take a look.

Sorry for the bad manners. 🙏🏻

shawnhwei commented:
Bumping as I've also encountered this issue many times over the years.

TylerHelmuth commented:
@mattfarina, @technosophos  as discussed in the SIG meeting call today, I spent some time working on tests for this PR. I first started investigating why the [existing test](https://github.com/helm/helm/blob/b307c08c2be41f68424484b4700a61358cfcceb0/pkg/chartutil/coalesce_test.go#L66) passes. That test has several `null` merging scenarios with global and subcharts so I wanted to understand why they are properly deleted/merged and how the test differs from an actual `install` or `template` scenario.

Debugging via a `template` command and test scenario, I found that in an actual install scenario the call to `chartutil.ProcessDependenciesWithMerge` [here](https://github.com/helm/helm/blob/b307c08c2be41f68424484b4700a61358cfcceb0/pkg/action/install.go#L249) ends up modifying the chart's `Values` (I believe on purpose, as part of `Imports`) to include any default values from subcharts.

The subchart's default values being present in the parent's values ends up causing problems during the second call to `CoalesceValues` in the [Render function](https://github.com/helm/helm/blob/b307c08c2be41f68424484b4700a61358cfcceb0/pkg/chartutil/values.go#L161). When the subchart's default values are present in the parent chart's values.yaml, they end up being used, as described in the PR description, during the `coalesceDeps` step of `coalesce`, after the key had already been deleted.

The current unit tests does not have this problem because it does not call `ProcessDependenciesWithMerge` and the setup does not include any subchart values in the parent's values. As a result, when `coalesceDeps` is called there is no key to add back.

I've updated the test to show this failure scenario [here](https://github.com/TylerHelmuth/helm/commit/c34cd0c254102b35139e157e485581e72a0e36ce) by manually adding a subchart's default value to a parent chart's values. I have also (not rigorously) confirmed [here](https://github.com/helm/helm/compare/main...TylerHelmuth:helm:subchart-null-bug) that the proposed changes in this PR fixes this scenario.

I am going to work on more test cases/scenarios and try to determine if the fix should be made in coalesce, as proposed by this PR, or somewhere in `ProcessDependenciesWithMerge`. My goal will be to create enough test scenarios that you feel confident that this PR does not create a Whack-a-Mole scenario and the change and its new tests can be merged (or some equivalent change in `ProcessDependenciesWithMerge` if that ends up being more appropriate).

ivankatliarchuk commented:
Is there is a way to add `--experimental or --beta` flag or similar, so that we could release it and community to test it? Another option is to use https://github.com/helm-unittest/helm-unittest on top of unit tests itself

jkroepke commented:
Welcome @gjenkins8 - the community invites you to solve this issue. 

Thank you.

xrow commented:
I need this bugfix to override security contexts in OCP and Gitlab. My current workaround is similar to this :-(.

find ./cp-gitlab -type f -name "values.yaml" -exec sed -r -i 's/runAsUser:(\s+\w*|)/runAsUser: null/g' {} +
find ./cp-gitlab -type f -name "values.yaml" -exec sed -r -i 's/fsGroup:(\s+\w*|)/fsGroup: null/g' {} +
find ./cp-gitlab -type f -name "values.yaml" -exec sed -r -i 's/runAsGroup:(\s+\w*|)/runAsGroup: null/g' {} +

rehevkor5 commented:
@TylerHelmuth I think you should make a PR against `ryanhockstad:subchart-null` which adds your unit tests, and get them to merge it? The addition of a meaningful unit test will make this PR more mature.

TylerHelmuth commented:
I have gotten pulled into other work and have not been able to get to the next step of creating tests. I believe to really prove this out we'd need some end-to-end tests that render some files to compare against expected renders. This ensures all the parts of the process are considered (not doing this is why the current unit test pass). The problem is there isn't any existing end-to-end tests like this to use, so it would mean creating a new process.

If someone wants to help move this forward the best thing to do next, in my opinion, is make a bunch of tests that go through the whole render process, as if you were doing a `helm template`, and compare the results to expected results.  I don't have time for this right now, but I hope someone else does.

ExNG commented:
When templating a chart with nested dependencies i've encountered a case of untyped subchart values.
Tested with patched helm `3.15.4` and `3.16.2`, also with unpatched `3.14.4`, `3.15.4` and `3.16.2`.

My test case was done using the [camunda/camunda-platform](https://github.com/camunda/camunda-platform-helm/blob/main/charts/camunda-platform-8.6/Chart.yaml) at v11.0.1, but interesstingly not with v10.4.2 and [bitnami/wordpress](https://github.com/bitnami/charts/blob/main/bitnami/wordpress/Chart.yaml) at v23.1.23 which have sub-sub-dependencies too.

The debug error printed is:
```
Error: template: test/charts/camunda/charts/identityKeycloak/templates/statefulset.yaml:163:27: executing "test/charts/camunda/charts/identityKeycloak/templates/statefulset.yaml" at <include "keycloak.databaseSecretName" .>: error calling include: template: test/charts/camunda/charts/identityKeycloak/templates/_helpers.tpl:168:75: executing "keycloak.databaseSecretName" at <.Values.postgresql.auth.existingSecret>: wrong type for value; expected string; got interface {}
helm.go:86: 2024-10-24 08:31:48.294951444 +0000 UTC m=+0.172370966 [debug] template: test/charts/camunda/charts/identityKeycloak/templates/statefulset.yaml:163:27: executing "test/charts/camunda/charts/identityKeycloak/templates/statefulset.yaml" at <include "keycloak.databaseSecretName" .>: error calling include: template: test/charts/camunda/charts/identityKeycloak/templates/_helpers.tpl:168:75: executing "keycloak.databaseSecretName" at <.Values.postgresql.auth.existingSecret>: wrong type for value; expected string; got interface {}
```

I did test other revisions of the camunda/camunda-platform v11 chart but didn't find a working state, this might hint on a verry rare case in which this fix might not work.
Luckily this specific chart includes a `global.compatibility.openshift.adaptSecurityContext` option as fallback.
I hope this insight is helpful, though i am sorry i cannot provide a fix.

ryanhockstad commented:
@ExNG I'm not sure what values you used in your test chart, but I think I know the problem.  If you are using the default `values.yaml`, it [explicitly sets](https://github.com/camunda/camunda-platform-helm/blob/10d6034c0cc900ecab0be9cbc0d98cb372d31051/charts/camunda-platform-8.6/values.yaml#L2597) the `identityKeycloak.postgresql.auth.existingSecret` (which is in the postgresql subchart to the keycloak subchart) to a null value. The whole premise of this PR is to respect that null value rather than have it get overwritten.  As you can see [here](https://github.com/bitnami/charts/blob/754f9e2365ab6d0786a997e4d582e2b6ce92c237/bitnami/keycloak/values.yaml#L1316), the Keycloak subchart does explicitly set this value to an empty string. 

So this raises a pretty existential question for this PR: Should this bug be patched if people are accidentally depending on it?

On one hand, I think it's definitely safe to call this a bug.  In the Camunda chart case, it may be very helpful that values they are explicitly setting end up getting overwritten, but that still seems like a bug.

On the other hand... this could be a disruptive patch for some existing helm charts.  I think the error you saw made it fairly clear what the problem was and should make it quick for chart maintainers to update their code.  It should also be easy enough for an end user to patch with their custom values file.  Having said that, Helm wants to be very backwards compatible, and in some cases, this will be a breaking change. I would understand if the maintainers want to move this to Helm 4 because of that.

I would obviously like it if it got merged before Helm V4, and I know that is the prevailing feeling here.  But I think we need to hear from a Helm maintainer on this.

MrBlaise commented:
> > On the other hand... this could be a disruptive patch for some existing helm charts
> 
> While I agree that this will now be disruptive for some, I'm not sure you're considering that this change has already been made and has already broken a bunch of people's Helm charts, and continues to do so as people upgrade their version of Helm to a "fixed/broken" version. Additionally, I'd bet that far less people are depending on this new logic compared to the entire history of Helm utilizing this. Finally, and more importantly, we, in Helm, NEED a way to unset a value in a subchart. Currently, this is impossible with the latest version of Helm. I personally stay out of date because of this. Subcharting in Helm is one of the more critical features if you ask me, it allows you to fully templatize and craft an entire (simple) application stack in a single helm chart with multiple sub-charts. Without a way to override a subchart, what you end up with is then needing to fork/clone the upstream chart and maintain it yourself to be able to change some "dumb" default that has no easy way to disable besides unsetting that value.
> 
> Overall, I think the good that undoing the bad that was done by "fixing" this outweighs the bad. By far. Unless you can come up with a way to unset/override subcharts with some other internal keyword meaning the same thing?

Exactly! We haven’t updated argocd since this issue occured, because we are refusing to rewrite our helm charts to the obviously wrong version. People who already changed their helm charts, will change it again and things will go back to normal. When this bug was introduced it caused a lot of headaches, and the longer it is not fixed the bigger headache it will cause again when it finally fixed, but that doesn’t mean it shouldn’t.

Just because we can workaround a bad design doesnt justify the bad design.

xrow commented:
Idea 1: If you are concerned about BC, maybe it is time for helm 2.X.X.
Idea 2: Have a similar field like kubeVersion called helmVersion in Chart.yaml

AndrewFarley commented:
Can someone just merge and release this already?  What's the blocker?  Whose door do I need to bust down?

fcrespofastly commented:
I have to say this issue is making charts dependency completely broken :(

MrBlaise commented:
Any updates here, what is the latest blocker?

xmj commented:
Reading through this thread I am reminded of https://xkcd.com/1172/.

What's the holdup in getting it in by now? One does lose track...

xrow commented:
@scottrigby @karenhchu @mattfarina Trying to poke a relevant maintainer. Whoever the proper person is. To get direction, guidance or management to stop the ongoing loop of ideas and comments. Please take a moment care. Ty.

pisto reviewed (approved):
As far as I can tell, this is ok.

karenhchu commented:
Hi all! Thank you for the contribution. The maintainers need to take a deeper look since there have been several attempts on this issue. That being said, as a next step to help this move along – can we start with getting some tests?

scottrigby commented:
Thanks everyone for your feedback and contributions on this.

I can understand some frustration here, but—as Karen said https://github.com/helm/helm/pull/12879#issuecomment-2551982738—PRs need tests in order to get merged. Especially when the PR affects code that could have larger consequences. And that's exactly the case here, since this is part of a larger issue with Helm values coalescing. Helm has because a critical piece of infrastructure tooling for many large organizations, and we need to make sure a fix for one issue doesn't break things elsewhere.

Note that was an earlier fix to the coalesce issue merged, but was later rolled back because it broke something else. This is because the coalesce logic in Helm is very confusing—the precedence in code was different in some cases to what was documented, and we didn't have tests that clearly defined this precedence order in the coalesce logic. Because of this, we held weekly meetings on this topic for a time with a goal of coming to a final resolution on the precedence question, and we have done that. The results are in this Google doc, which also links to a number of issues and pull requests. Please take a look: https://docs.google.com/document/d/1ZjmIc5VYzNbaJ-SQ6nT02Oo1t9ctlvomufP8ulR6q70/edit?tab=t.0#heading=h.stcvhj9ydagu

I would like to help address these connected issues properly now that we have resolved the precedence discrepancy. Several other Helm maintainers would too. Will add a link here to new issue/PR for that.

scottrigby committed:
Tests for bugfix: Override subcharts with null values #12879

- Add consistency for null test in given values, parent chart, subchart, and
  sub-sub-chart
- Remove bar null test to keep consistent with boat=null at top level

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby commented:
Happy holidays everyone ❄️🌲

I added some tests as a linked git patch, and based on that believe this PR works as expected without breaking other things—at least what we have test coverage for and I could think to test manually.

This patch can be git applied and pushed to this PR branch by @ryanhockstad, or a Helm maintainer can do it. 

I am happy to do that, but I would like some community feedback first. I believe the use cases for null values are now covered in these tests via this patch, but please take a look and give feedback about whether I missed any relevant use cases here.

Note that this patch also:
- Adds consistency for null test in given values, parent chart, subchart, and    
  sub-sub-chart                                                                 
- Removes bar null test to keep consistent with boat=null at top level

If you want to test locally:

1. checkout this PR branch locally
  _for folks unfamiliar with how to do this, run `gh pr checkout 12879` in your local clone, using the [GitHub CLI](https://cli.github.com/)_
2. download this patch file [0001-Tests-for-bugfix-Override-subcharts-with-null-values.patch](https://github.com/user-attachments/files/18246595/0001-Tests-for-bugfix-Override-subcharts-with-null-values.patch)
3. run `git apply /PATH/TO/PATCHFILE`
4. run `go test ./pkg/chartutil -run TestCoalesceValues`. You should get `ok  	helm.sh/helm/v3/pkg/chartutil`

scottrigby commented:
Fixed milestone. This will also need a v3 backport (at this point, it should be the exact same commits)

ryanhockstad committed:
Add test case for removing an entire object

Signed-off-by: Ryan Hockstad <ryanhockstad@gmail.com>

ryanhockstad commented:
Thanks for adding these tests, @scottrigby! I can confirm they are failing in the current `main` branch, but passing in this PR, so I added your commit.

As far as coverage goes, I think we should add a test that 'nullifies' an entire map.  I've added that change in [this commit](https://github.com/helm/helm/pull/12879/commits/ef2eb552837446fe8a5a91bc68abf4d9eab685a8). What do you think?

scottrigby commented:
> Thanks for adding these tests, @scottrigby! I can confirm they are failing in the current `main` branch, but passing in this PR, so I added your commit.

👍

> As far as coverage goes, I think we should add a test that 'nullifies' an entire map.  I've added that change in [this commit](https://github.com/helm/helm/pull/12879/commits/ef2eb552837446fe8a5a91bc68abf4d9eab685a8). What do you think?  

I think it's a good addition 👍

ryanhockstad committed:
add test for nullifying nested global value

Signed-off-by: Ryan Hockstad <ryanhockstad@gmail.com>

ryanhockstad commented:
@TylerHelmuth  I think adding a test for global is a good idea to address how that section is handled differently.  I added a test in the latest commit [here](https://github.com/helm/helm/pull/12879/commits/8da0374896fabdec12b0d9ab3ef6f523c001448f).  Setting the new global `foo2` to null under the `pequod` section nullifies its `ahab` subchart global value as expected.

I think the reason this works without requiring the code change you suggested is because the bug that my change fixes prevents `coalesce` from forgetting about explicit `null`s set in a subchart.  The logic that handles dealing with `global` values is totally separate.

ryanhockstad commented:
I was just testing this out a little more locally, and there is a weird case where one chart might set a default for a global value and a subchart could set a default for the same value. 

In this case, if a user were to explicitly set the top level global to `null`, the subchart's global value would not be nullified. This is because the top level global would disappear, so nothing at the top level would exist to nullify the default subchart global value.

@scottrigby if that is the situation you were thinking about, I think the best solution would probably be for chart users to explicitly nullify the global value at the top level and global value within the subchart, which works.  I could update the test I just added to reflect that.

V02460 commented:
First, thank you for working on this PR! I have some concerns with it, though: While it fixes important issues, it is another bandaid on top of the more fundamental issue of coalescing: Overloading the meaning of `null` will always bite someone in some way.

A better solution, to my understanding, would be to completely replace coalescing with simple merging. This will make it easy to understand how final values are computed and eliminates all edge cases that come with the current inconsistent mix of both. Remove coalescing and there is no more double meaning of what `null` means.

For the rare cases where a key with a `null` value is interpreted differently to a non-existing key (a bad anti-pattern in my opinion), the considered `--unset` command line flag could be realized. Or, more broadly, a patch mechanism like, e.g., JSON Patch could support key deletion in a clean way besides offering additional handy operations. Whatever the tool to perform this deletion, it must be available, but be separated from the basic merging of Helm values.

This PR is a move in the right direction by removing some uses of coalescing, but I think the proper solution is abolishing it completely. Thanks for considering!

scottrigby commented:
@V02460 I agree this is something we should reconsider for Helm 4. Now `main` is for Helm 4, while Helm 3 backports should be made against `dev-v3` branch after they are fixed in `main`.  That said, this PR is _very_ close to being ready to merge. At the moment a back port would be nearly as-is—because most of the code in both branches is the same, although I expect this to change very soon as we begin to make major architectural changes for Helm 4. Because of that, I would like to see this fixed in `main` with the current functionality—if possible—before making a change to the method we use to delete keys. Note that I initially introduced the functionality in Helm to delete keys, and because of increasing complexity of coalescing over time and problems caused, I suggested the `--unset` alternative and still think that would be a better solution in Helm 4. That would be a separate PR, and should start as a [Helm Improvement Proposal (HIP)](https://github.com/helm/community/tree/main/hips).

scottrigby commented:
> I was just testing this out a little more locally, and there is a weird case where one chart might set a default for a global value and a subchart could set a default for the same value.
> 
> In this case, if a user were to only explicitly set the top level global to `null`, the subchart's global value would not be nullified. This is because the top level global would disappear, so nothing at the top level would exist to nullify the default subchart global value. Likewise, if a user only nullified the subchart global value, it would end up still getting overwritten by the top level default global value.
> 
> @scottrigby if that is the situation you were thinking about, I think the best solution would probably be for chart users to explicitly nullify the global value at the top level and global value within the subchart, which works. I could update the test I just added to reflect that.

@ryanhockstad I have an idea. How about waiting until after coalescing is finished to delete any keys that have a remaining (non-overridden) `null` value? That way, existing coalescing precedence would be respected.

ryanhockstad commented:
@scottrigby I'm not sure the added complexity of adding logic to nullify certain keys after the merge/coalesce recursion has occurred once is worth dealing with edge cases that users are able to handle themselves anyway.

Also, the way that this should be handled is a bit ambiguous anyway.  A user may want to actually nullify a parent global key without nullifying all of the children globals.

scottrigby commented:
OK so I think we should keep the scope of this PR to a fix for known issues users are having. It addresses them very well with many new tests, and as such is a valuable improvement. Let's move consideration for fixing possible issues with globals and null values to a separate issue. I'm running the tests now—let's see where they land.

scottrigby reviewed (approved):
lgtm 👍

joejulian reviewed (approved):
lgtm

scottrigby commented:
Thanks @joejulian and everyone 👏 Merging

@ryanhockstad do you want to open a v3 backport PR from your same branch while the code applies as-is, and link to this one?

AndrewFarley commented:
10 months later, amazing, thank you team.  Now just waiting for the backport and release of Helm 3 with this patch

ryanhockstad commented:
Here is the backport PR: https://github.com/helm/helm/pull/13654

pierluigilenoci commented:
@ryanhockstad thank you!

scottrigby commented:
📣 Backport PR https://github.com/helm/helm/pull/13654 merged 👏 This fix is scheduled to be in Helm 3.17.1
